### PR TITLE
optimize: rewrite while loops as for loops

### PR DESCRIPTION
**Pull request by dlax**
I find it clearer this way and this avoids infinite loops or maximum iterations exceed that sometimes occurs.
For now, this is only done in Newton-CG method but I can do it for other methods as well if needed.

### CODE REVIEW

**Comment by ev-br** [comment]
I'm +0 here. Will hit the green button before branching 0.18.0 unless further comments.

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -1431,24 +1431,43 @@ def _minimize_newtoncg(fun, x0, args=(), jac=None, hess=None, hessp=None,
...
     retall = return_all
 
+    def terminate(warnflag, msg):
+        if disp:
+            print(msg)
+            print("         Current function value: %f" % old_fval)
+            print("         Iterations: %d" % k)
+            print("         Function evaluations: %d" % fcalls[0])
+            print("         Gradient evaluations: %d" % gcalls[0])
+            print("         Hessian evaluations: %d" % hcalls)
```
Comment (perimosocordiae) [inline scipy/optimize/optimize.py]: Nitpick: seems simpler to write

    print("         Hessian evaluations:", hcalls)

for each of these.

**Comment by perimosocordiae** [comment]
+1 here, the control flow reads much more easily now.

**Comment by samtygier** [comment]
This would break the case that the user set maxiter=np.inf (i.e. never stop unless converged). May not be a real problem, but this is handled for the methods that have both maxiter and maxfev as parameters (for these it is more useful as maxiter=np.inf explicitly means only consider the maxfev limit).

**Comment by Sam-Tygier** [comment]
Also breaks if the user passes the limit as a float. I often write `1e4` instead of `10000` in places that I can get away with it.

Should probably be xrange() (with a from scipy._lib.six import xrange ) , in case someone sets maxiter to a very large int on python 2.

**Comment by samtygier** [comment]
Also breaks if the user passes the limit as a float. I often write 1e4 instead of 10000 in places that I can get away with it.

Should probably be xrange() (with a from scipy._lib.six import xrange ) , in case someone sets maxiter to a very large int on python 2.

**Comment by dlax** [comment]
samtygier a écrit :
> This would break the case that the user set maxiter=np.inf (i.e. never
> stop unless converged). May not be a real problem, but this is handled
> for the methods that have both maxiter and maxfev as parameters (for
> these it is more useful as maxiter=np.inf explicitly means only consider
> the maxfev limit).
>

I was not aware of this kind of tricks. In that case, I guess using a 
for loop would only be applicable for the inner loop which `cg_maxiter` 
is internally defined.

**Comment by ev-br** [comment]
Since this is not a release blocker, I'm carrying the milestone forward.

**Comment by perimosocordiae** [comment]
It seems easy enough to fix this without ditching either for loop:
 1. use `xrange`, as @samtygier suggested
 2. coerce `maxiter` to `int`, possibly like this:

  ```python
  try:
      maxiter = int(maxiter)
  except OverflowError:  # triggered by infinity
      maxiter = int(1e100)
   ```

**Comment by samtygier** [comment]
In python2
```
>>> xrange(int(1e100))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: Python int too large to convert to C long
```
`xrange(sys.maxint)` looks like the biggest you can do (python3 seems not to have the restriction). Thats fine on 64bit (a quick check says it would take a few thousand years just to run an empty loop), but on 32bit (I only checked on arm) sys.maxint is only 2**31-1. You'll need to use MAXSIZE from six so that it works on py2 and py3.

I'd guess that 2**31-1 is probably close enough to inf to keep people happy, especially as I don't imagine many people are doing big work on 32bit machines any more.

**Comment by dlax** [comment]
Just pushed a version with only the inner loop as a for loop. At least factoring out the termination seems like an improvement and the remaining while loop doesn't look that bad. Not sure it's worth playing with maxint here.

**Comment by ev-br** [comment]
Okay, no more comments for a week, let's merge this. Thanks @dlax @perimosocordiae @samtygier

**Pull request merged**